# Heurísticas de creación de algoritmos

En general, tenemos dos tipos de problemas algorítmicos.: 

- **Optimización.** Queremos encontrar un mínimo o un máximo para un problema en cierto **espacio de estados**
- **Decisión.** Queremos responder de sí/no también para un problema en cierto espacio de estados

No todos los problemas de algoritmos son así, pero una cantidad muy grande de los que son intersantes sí. Por ejemplo, insertar un elemento en una lista no está intentando optimizar ni decidir nada. Pero hay algunos problemas que ya hemos vito de este estilo. El problema de buscar un elemento en un diccionario es un problema de decisión: dado un diccionario $D$ y un elemento $x$, lo que queremos responder es si $x$ está en $D$. 

Veamos algunos ejemplos más: 

- **Resolver un Sudoku.** Dadas entradas en un tablero de Sudoku, decidir si hay una solución que lo complete o no. También, decidir si esta solución es única o no. 
- **Brazo robótico que une circuitos.** Dados puntos en el plano, queremos encontrar un ciclo hamiltoniano de longitud mínima que los recorre. 
- **Decidir el número cromático de una gráfica.** Dada una gráfica $G$ queremos encontrar la mínima cantidad posible de colores necesarios para poder dar una buena coloración.
- **Determinar el número de clique de una gráfica.** Dada una gráfica $G$ queremos encontrar la máxima cantidad posible de vértices que forman una gráfica completa. 
- **Determinar si una gráfica es bipartita o no.** Dada una gráfica $G$, ver si existe una partición de sus vértices en conjuntos $A$ y $B$ de mode que las únicas aristas vayan de $A$ a $B$. 

El primero y último son algoritmos de decisión, mientras que los demás son algoritmos de optimización. 

## Espacio de estados
En un problema de decisión o de optimización, es muy importante que quede claro el **espacio de estados**, es decir, todas las posibles configuraciones/entradas que debemos considerar para poder resonder la pregunta. Para ello, en problemas de aplicación es muy importante decidir cómo estamos modelando el problema. 

Espacios de estados típicos en varios de estos problemas son: 

- Todas las permutaciones de $n$ elementos
- Todos los conjuntos de $n$ elementos
- Todas las configuraciones de $n$ puntos en el plano
- Todas los números del $1$ al $n$
- Para cierta $k$, todos subconjuntos de tamaño $k$ de $n$ elementos
- Todos los vectores de $m$ elementos tomados de un conjunto de $n$ elementos

**Ejemplo.** Dada una lista de $n$ números, queremos:

- Decidir si hay dos de ellos cuya suma es $1000$
- Decidir cuáles dos de ellos tienen la suma más pequeña

El primer problema es un problema de decisión. El segundo es un problema  de optimización. Notemos que ambos problemas tienen como espacio de estados a los subconjuntos de $2$ elementos de un cocnjunto de $n$ elementos. 

Hay otros problemas que tienen espaciones de estados más complicados, o m´sa particulares al problema. Por ejemplo, consideremos los siguientes dos problemas: 

**Ejemplo.** 

- ¿Será posible colocar 15 caballos de ajedrez en un problema sin que se ataquen entre sí?
- ¿Cuál es el máximo número de caballos de ajedrez que se pueden poner en un tablero de ajedrez sin que se ataquen entre sí?
- ¿Será posible colocar $3$ torres, $5$ caballos y  $4$ alfiles sin que se ataquen entre sí? 


# Heurísticas algorítmicas

Ya que tenemos un problema algortítmico de decisión o de optimización y entendemos bien cuál es el espacio de estados que debemos estudiar, lo siguiente es saber dónde en ese espacio de estados se encuentra la solución óptima o bien, la instancia que cumple lo que queremos. 

Hay muchas formas de resolver este tipo de problemas algorítmicos, pero en transcurso de la historia del análisis de algoritmos, estas formas se han agrupado en **heurísticas** generales que ayudan en muchas situaciones.

A continuación ponemos algunas:

- Explorar todo el esapcio de estados (fuerza bruta): consiste en estudiar todos los elementos del espacio de estados uno por uno para ver si son el óptimo/cumplen la propiedad que queremos.
- Explorar el espacio de estados de manera inteligente: consiste en estudiar parcialmente el espacio de estados, descartando con suficiente anticipación las exploraciones que ya no serán exitosas. 
- Explorar el espacio de estados de manera voraz (greedy)
- Reducir el espacio de estados con argumentos de simetría
- Dividir el problema que queremos en problemas más pequeños que sean más sencillos de resolver. 
- Explotar una estructura recursiva de los objetos del problema para poner soluciones a instancias grandes en términos de soluciones de instancias más pequeñas
- Programación dinámica: hacer lo anterior con mucho más cuidado para no repetir múltiples veces el cómputo para casos pequeños

## Exploración exhaustiva
Consiste en explorar todo el espacio de estados para buscar el valor óptimo o el testigo. Es una técnica básica, pero que a veces es la única con la que cocntamos. Usualmente es la única opción en problemas con muy poca estructura, o en probelams en donde queremos asegurarnso de pasar por todas las posibilidades. 

También se le conoce como "fuerza bruta", o como "explorar por completo el espacio de estados". 

### Problema 1
¿De cuántas formas se pueden poner a $10000$ como suma de cuadrados de dos números enteros positivos? ¿En cuál de las expresiones $x^2 + y^2 = 10000$ se minimiza $3x + 5y -1$. 

Para el Problema 1, el espacio de estados que queremos explorar son las parejas $(x,y)$ con $x$ y $y$ en el intervalo $[1,99]$. Una exploración exhaustiva verifica todos los casos posbiles. Haremos esto en Python haciendo dos ciclos.

In [5]:
# Primero, la exporación exhaustiva para ver quienes son todas las parejas
cuantos = 0
cuales = []
for x in range(1,100):
    for y in range(1,100):
        if x**2 + y**2 == 10000:
            cuantos += 1
            cuales += [(x,y)]
print(cuantos)
print(cuales)

# Ahora, hagamos otra exploración para ver en qué pareja se minimiza 3x+5y-1
minimo = 100000000
for pair in cuales:
    x = pair[0]
    y = pair[1]
    if 3*x + (5*y) -1:
        minimo = 3*x + (5*y) - 1
        optimo = (x, y)

print(minimo)
print(optimo)

4
[(28, 96), (60, 80), (80, 60), (96, 28)]
427
(96, 28)


Pensemos que el Problema 1 se generaliza para dos números $x$ y $y$ que queremos que su cuadrado sume $n$. En este caso, el espacio de estados sería, de acuerdo a nuestra estrategia anterior, que $x$ y $y$ estén en ${1,2, \ldots, \lceil \sqrt{n} \rceil}$

En el primer ciclo estamos ccorriend por $O(\sqrt{n})$ elementos y el que está anidado también corre por $O(\sqrt{n})$ así que estos ciclos anidados corren en tiempo $O(n)$. Con este tiempo se puede tanto determinar cuáles parejas son, como determinar cuál es el mínimo de la expresión $3x+5y-1$ sujeta a las condiciones $x^2 + y^2 =n$ y $x,y$  enteros.

### Problema 2

Se tomas tres enteros $a$, $b$ y $c$ en el intervalo $[1, 100 ]$. ¿Para cuáles de ellos al valor de $a^2 + 2b^2 + 3c^2 - 2ab -5bc -7ca$ es mínimo?

In [9]:
def funcion(a, b, c):
    return (a**2 + 2*b**2 + 3*c**2-2*a*b-5*b*c-7*c*a)

minimo = 10000000
for a in range(1,101):
    for b in range(1, 101):
        for c in range(1, 101):
            F = funcion(a,b,c)
            if F < minimo:
                minimo = F
                optimo = (a,b,c)

print(minimo)
print(optimo)

-80000
(100, 100, 100)


Este problema se presta mucho a una exploración exhaustiva total, pues no hay tanta simetría en las variables $x,y,z$. Si el problema fuera para números en el intervalo $[1,\ldots,n]$ entonces el algoritmo de acá arriba ccorre en tiempo $O(n^3)$.

### Problema 3.
¿Cuál es la palabra más larga en español que tenga únicamente cuatro vocales? 

Tomaremos como espacio de estados la lista en <a href="http://www.gwicks.net/dictionaries.html"> este sitio </a>

Vamos a hacer una exploración exhaustiva palabra por palabra para determinar cuáles tienen exactamente cuatro palabras y ver cuáles de ellas es la más grande. 


In [2]:
vocales = 'aeiouáéíóúAEIOUÁÉÍÓÚüÜ'
def contarvocales(palabra):
    cuenta = 0
    for j in palabra:
        if j in vocales:
            cuenta += 1
    return cuenta

print(contarvocales('Hola mundo!'))
print(contarvocales('Esta oración tiene acentos'))
print(contarvocales('PIngüinos y MurCIELAgos'))

4
12
9


In [16]:
vocales = 'aeiouáéíóúAEIOUÁÉÍÓÚüÜ'
lista = open('espanol.txt', 'r', encoding='ISO-8859-1')
linea = lista.readline()
maximo = 0
while linea:
    limpio = linea.split(' ')[0]
    if contarvocales(limpio) == 4:
        if len(linea) > maximo:
            maximo = len(limpio)
            mejor = limpio
    linea = lista.readline()

print(maximo)
print(mejor)

13
yuxtapondrán



**Problema 4.** ¿Existe alguna palabra en español que use exactamente diez vocales y diez consonantes? Si sí, ¿cuántas hay?

In [7]:
lista = open('espanol.txt', encoding='ISO-8859-1')
linea = lista.readline()
testigos = []
vocs = 10
consonantes = 20
while linea:
    limpio = linea[:-1].split(' ')[0]
    if contarvocales(limpio) == vocs and len(limpio) == consonantes:
        testigos.append(limpio)
    linea = lista.readline()

print(f'Hay {len(testigos)} palabras con {vocs} vocales y {consonantes} consonantes.')
for t in testigos:
    print(t)
lista.close()

Hay 2 palabras con 10 vocales y 20 consonantes.
desnacionalizaciones
impermeabilizaríamos


Consideremos el siguiente problema que se parece al anterior, pero en vez de ser de decisión es de optimización:

**Problema.** Encontrar el mayor entero $k$ tal que en español hay una palabra que use exactamente $k$ vocales y $k$ consonantes. 

Observemos que podemos resolver este problema de optimización resolviendo varias instancias del siguiente problema de decisión:

**Problema.** Dado un entero $k$, determinar si en español hay una palabra que use exactamente $k$ vocales y $k$ consonantes.
 
 Aunque a veces este estrategia es lo mejor para algunos problemas, hay otros problemas en los que hay mejores formas de resolver la versión de optimización. 


**Problema 5.** Las letras $a,b,c,d,e,f,g, h, i, j$ representan dígitos distintos, ¿para cuántas elecciones tenemos que $\frac{abcde}{fghij}$ es un número entero? Aquí $abcde$ y $fghij$ son los números de cinco dígitos obtenidos de concatenar los dígitos correspondientes. 

Aquí hay que decidir cuidadosamente el espacio de estados. Si elegimos como espacio de estados que cada letra tome cada una de las 10 posibilidades que tiene, y luego verificamos duplicados y luego procesamos, se tendrán que verificar $\approx 10^{10}$ casos. Esto es demasiado, pues son 10 mil millones de casos. 

¿Queremos que sean todas las permutaciones posibles? Son $10!=3628800$. Está bien, no son tantísimas. Pero este sería un tiempo imposible para permutaciones de más números. 

In [2]:
def perms(lista):
    L = len(lista)
    if L == 1: 
        return [lista]
    else:
        old = perms(lista[:-1])
        last = lista[-1]
        new = []
        for k in range(L):
            for perm in old:
                to_add = perm[:k] + [last] + perm[k:]
                new.append(to_add)
        return new

#numeros = list(range(1, 11))
#permutaciones = perms(numeros)
#print(len(permutaciones))

In [13]:
soluciones = []
for pi in permutaciones:
    first = [str(j) for j in pi[:5]]
    last = [str(j) for j in pi[5:]]
    a = int(''.join(first))
    b = int(''.join(last))
    if a % b == 0:
        soluciones.append((a,b))
    
print(f'Hay {len(soluciones)} soluciones')

Hay 21 soluciones


## Exploración exhaustiva recortada

Consiste en dejar de explorar posibilidades que ya no se pueden extender a posibilidades exitosas. Esto no es lo miso que reducir el espacio de estados. Tampoco quiere decir que no exploremos todas las posibilidades. Consiste en dar argumetnos para que nuestro algoritmo revise menos casos y siga dando la respuesta correcta. 

Retomemos uno de los problemas de la sección anterior:

### Problema 1
¿De cuántas formas se pueden poner a $10000$ como suma de cuadrados de dos números enteros positivos? ¿En cuál de las expresiones $x^2 + y^2 = 10000$ se minimiza $3x + 5y -1$. 

Recortemos el espacio de estados a partir de las siguientes dos observaciones:
- Ambos $x$ y $y$ tiene que se pares
- Si $x$ ya está dada, no tiene chiste mover $y$ por todos sus valores posibles de $1$ a $100$ pues por tamaño ya solo puede tenre pocas posibilidades.

$y$ ya solo puede ser a lo mucho $\approx{\sqrt{10000 - x^2}}$. Esto ahorra algunos pasos- 

Los otros problemas 

In [14]:
soluciones = []
for x in range(1, 101):
    for y in range(1, int((10000-x**2)**(0.5)+1)):
        if x**2 + y**2 == 10000:
            soluciones.append((x,y))
print(soluciones)

[(28, 96), (60, 80), (80, 60), (96, 28)]


## Problemas de exploración exhaustiva recortada

Pongamos otros cuantos problemas que se pueden estudiar usando un recorte de espacio de estados. 

**Problema 6.** Tenemos que encontrar todas la parejas de palabras $x$ y $y$ en español que sean diferentes y que $x y$ tenga en total 9 caracteres. 

**Problema 7.** Queremos encontrar para cuantas parejas $x$ y $y$ de palabras en español sucede que cada una de ellas tiene por lo menos cinco letras y las últimas cinco letras de la primera son iguales a las primeras cinco letras de la última. 

**Problema 8.** Tenemos el sigiuente arreglo de números $$[4,1,7,4,2,6,5,7,1,8,4,9,1,4,1,5,7,2,8,3,6]$$
 Queremos determinar de cuántas formas se pueden elegir algunos de estos números de forma consecutiva de modeo que sumen $18$.

 **Problema 9.** Un 

**Problema 6.** Tenemos que encontrar todas la parejas de palabras $x$ y $y$ en español que sean diferentes y que $x \quad y$ tenga en total 9 caracteres. 

Podemos pensar el espacio de estados como todas las parejas de palabras en español y procesar cada una de ellas. Si lo exploramos de manera directa esto toma tiempo cuadrático en la cantidad de palabras en español. Al procesar cada posibilidad seguro que cubrimos todas, pero parece que hay cierta pérdida de tiempo pues las palabras grandes las estamos considerando en muchas parejas que van a fallar, es decir, las estamos considerando de manera innecesaria (palabras grandes tienen más de 7 letras). 

Mejor, podemos usar tiempo lineal en descartar las palabras grandes y luego usar itempo cuadrático en una lista mucho más corta. 


In [16]:
lista = open('espanol.txt', 'r', encoding='ISO-8859-1')
linea = lista.readline()
cortas = []
while linea:
    limpio = linea[:-1].split(' ')[0]
    if len(limpio) <= 7:
        cortas.append(limpio)
    linea = lista.readline()
print(len(cortas))

32974


Ya tenemos la lista de palabras cortas. Con esta lista ahora sí podríamos hacer una exploración exhaustiva. Como hay 1024 millones de casos a verificar, esto tardaría $\approx 1024$ en terminar. ¿Cómo podemos seguir reduciendo el espacio de estados? Agreguemos una idea de ordenar la lista de palabras por su longitud

In [19]:
cortas = sorted(cortas, key=lambda x: len(x))
print(cortas[:70])

# A continuación se muestra una búsqueda recortada
soluciones = 0
for j in cortas:
    for k in cortas:
        if len(j) + len(k) == 8: #no es 9 porque necesitamos contar el espacio vacío entre las dos palabras
            soluciones += 1
        if len(j) + len(k) > 8:
            break
print(soluciones)

['', '', '', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'ñ', 'al', 'as', 'ay', 'be', 'ce', 'ch', 'cu', 'da', 'de', 'dé', 'di', 'el', 'en', 'es', 'ex', 'fb', 'fe', 'fu', 'ha', 'he', 'ii', 'in', 'ir', 'iv', 'ja', 'je', 'la', 'le', 'lo', 'me', 'mg', 'mi', 'mí', 'ml', 'ni', 'no', 'oh', 'oí', 'os', 'pe']
7099844


Acá arriba recortamos la búsqueda en cuanto estuvimos seguros de que ya no habría soluciones. Esto nos dejó todavía con un algoritmo potencialmente cuadrático en $O(n^2)$ pero con un mejor factor constante que permitió correrlo en pocos segundos.

**Problema 9.** Una **matriz mágica sencilla** consiste de una matriz de $3 \times 3$, la suma de las entradas en cada fila es un cierto número $x$ y la suma de las entradas en cada columna es es ese mismo número $x$. Las entradas deben ser los números del $1$ al $9$ sin repetir. ¿Cuántas matrices mágicas existen?

Lo primero que tenemos que hacer es decidir quién será nuestro espacio de estados. Como los números no se repiten, podemos pensar en permutaciones. Esto en total nos da un espacio de estados de 9! elementos a verificar. Para hacer esto, tomamos la función auxiliar de permutaciones que hicimos arriba

La permutacion $[a_1, \ldots, a_9]$ la pensaremos como la matriz
$$\begin{pmatrix} a_1 & a_2 & a_3 \\ a_4 & a_5 & a_6 \\ a_7 & a_8 & a_9 \end{pmatrix}$$

In [21]:
numeros = list(range(1,11))
permutaciones = perms(numeros)
print(len(permutaciones))

3628800


Observación: la suma de todos los números del 1 al 9 es $45$. De este modo, la suma e ncada fila y en cada colmna debe ser igual a $15$. Esto nos permite comparar la suma de cada fila y columna no entre sí, sino entre ellas y un número constante. Esto nos permite poner la conddición de que la matriz sea mágina a ciertas sumas igualadas a 15.

In [26]:
soluciones = 0
for j in permutaciones:
    #si pasan las sumas que queremos
    soluciones += 1
print(soluciones)

3628800


## Recortes del espacio de estados con simetrías

Es usar simetrías en el espacio de estados y el problema planteado para reducir la exploración. Por ejemplo, si estamos buscando cuántos parejas $(x,y)$ de una lista de números suman cierto número $M$, entonces una posibilidad es que recorramos todo el espacio de estados para $x$ y $y$, pero otroa forma de hacerlo es suponer que $x \leq y$ y luego tras encontrar todas las soluciones bajo esta suposición, simplemente a partir de cada pareja $(x,y)$ construir la pareja $(y,x)$. 

Apliquemos esta idea al siguiente problema

**Problema 10.** Se cayeron los $12$ números de un reloj. Se quieren volver a poner, uno en cada posición. No importa tanto saber la hora, pero es muy importante que la suma de tres de esos números consecutivos (en orden cíclico) no sea $13$, porque da mala suerte. ¿De cuántas formas es posible hacer esto? 

Si hiciéramos una solución que explore todos los posibles estados, entonces necesitaríamos pasar por $12!$ de ellos, que son como $480$ millones. Sin embargo, podemos ahorrarnos un factor de $12$ si observamos que tenemos una simetría rotacional de orden $12$: una acomodo funciona si y solo si funcionan todas sus rotaciones. Por esta razón, basta con considerar aquellas permutaciones que comiencen con $1$ y verificar lo que queremos.  

In [5]:
numeros = [2,3,4,5,6,7,8,9,10]
candidatas = [[1] + j for j in perms(numeros)]
print(len(permutaciones))

3628800


In [7]:
candidatas[:5]

[[1, 10, 9, 8, 7, 6, 5, 4, 3, 2],
 [1, 10, 9, 8, 7, 6, 5, 4, 2, 3],
 [1, 10, 9, 8, 7, 6, 5, 3, 4, 2],
 [1, 10, 9, 8, 7, 6, 5, 2, 4, 3],
 [1, 10, 9, 8, 7, 6, 5, 3, 2, 4]]

In [9]:
def buena_suerte(lista):
    buena = True
    for j in range(10):
        buena = buena and (lista[(j%10)] + lista[(j+1)%10] + lista[(j+2)%10] != 13)
    return buena

soluciones = 0
for lista in candidatas:
    if buena_suerte(lista):
        soluciones += 1
print(soluciones)

165520


Tenemos $165520$ soluciones que comienzan con $1$, si queremos regresar al problema con $10$ números, todavía cada una de estas soluciones crea $10$ soluciones, una por cada una de las $10$ rotaciones posibles. Así, la cantidad total de soluciones es $1655200$. 

## Algoritmos voraces (greedy, glotones, ambiciosos)
Consisten en intentar realizar 

Comenzamos con el siguiente problema 

**Problema 11.** Se tienen números enteros positivos distintos $a_1, a_2, \ldots, a_n$. Se quiere encontrar una permutación de ellos $b_1, b_2, \ldots, b_n$ que haga la siguieten suma $$b_1+2b_2+3b_3+\ldots+nb_n$$ máxima. Resolver el problema en general y luego ejecutarlo para: 
```
50,76,72,52,70,74,84,16,43,29,31,77,22,92,46,38,91,42,98,87,83,6,44,7,36,80,11,10,82,67,90,18,27,37,86,33,64,26,9,48
``` 

En este problema el espacio de estados son todas las posibles permutaciones de $a_1, \ldots, a_n$, que son $n!$. Por lo tanto, si queremos resolver el problema por exploración exhaustiva, tendríamos que verificar $\Theta(n!)$ casos. Incluso cuando $n$ no es tan grande, esto es demasiado tiempo. 


Para resolver el problema, vamos a realizar un algoritmo voraz eligiendo los $a_i$ más grandes para los coeficientes más grandes. Es decir, el máximo de los $a_i$ va a acompañar al coeficiente $n$, el siguiente al coeficiente $n-1$ y así sucesivamente. 

Con esta idea se puede proponer el siguiente algoritmo: 

- Poner los $a_i$ en una lista de Python - Tiempo $O(n)$
- Tomar el $a_i$ más grande y declararlo como $b_n$ - Tiempo $O(n)$
- Eliminar ese $a_i$ de la lista - Tiempo $O(n)$
- Repetir para ir definiendo $b_{n-1},\ldots,b_1$ - En total tiempo $O(n^2)$

Este algoritmo toma $O(n^2)$ tiempo en total. Se puede mejorar a tiempo $O(n\log n)$ si primero ordenamos la lista y vamos tomando los más grandes. 


In [9]:
A = [50,76,72,52,70,74,84,16,43,29,31,77,22,92,46,38,91,42,98,87,83,6,44,7,36,80,11,10,82,67,90,18,27,37,86,33,64,26,9,48]
A.sort()
print(A)
print(sum(i*a for i,a in enumerate(A)))

[6, 7, 9, 10, 11, 16, 18, 22, 26, 27, 29, 31, 33, 36, 37, 38, 42, 43, 44, 46, 48, 50, 52, 64, 67, 70, 72, 74, 76, 77, 80, 82, 83, 84, 86, 87, 90, 91, 92, 98]
53023


Vamos a mostrar que la permutación de $a_1, \ldots, a_n$ que maximiza $a_i+2a_2+\ldots+na_n$ es precisamente en la cual tenemos los $a_i$'s ordenados. 

Pensemos que tenemos una permutación $b_1, b_2, \ldots, b_n$ en la cual los $b_i$ no están ordenados en orden creciente. Como los $b_i$ no están en orden creciente, debe existir una pareja de índices $i$ y $j$ tales que $i<j$ y $b_i>b_j$. Entonces la suma se ve así: 

$S=\ldots + ib_i + \ldots + jb_j + \ldots$

Al pasar $b_i$ a la posición $j$ y $b_j$ a la posición $b_i$, la nueva suma es:

$$S-ib_i -jb_j + jb_i + ib_j = S + (j-i)(b_i-b_j) > S$$

por lo cual dicha permutación que habíamos tomado no puede ser la mayor.

**Problema 12.** Se tienen $2n$ números, se quieren encontrar $n$ números tal que su suma sea meanor. 

```
34,-51,-67,-42,-72,27,-12,55,-69,47,89,-54,92,8,77,82,-52,-48,-61,-90,95,-49,32,18,40,96,65,44,81,24,38,-4,-14,-97,-20,63,86,-98,-19,-8
```

Una forma de resolverlo es con exploración exhaustiva sobre el espacio de estados de subconjuntos de $n$ elementos de un conjunto con $2n$ elementos.  Para cada uno de estos subconjuntos hacemos su suma y buscamos entre ellas la menor. Con este algoritmo tendríamos que verificar $\binom{2n}{n} casos. esto son $\Omega(2^{2n})$ casos, que como es exponencial tampoco nos conviene mucho.

Una mucho mejor forma es tomar de entre l alista los $n$ más pequeños. Estos definitivamente tienen la suma más pequeña. Tarda $O(n\log n)$ pues vamos a ordenarlos primero y despues tomar los primeros $n$ y sumarlos.

In [10]:
A = [34,-51,-67,-42,-72,27,-12,55,-69,47,89,-54,92,8,77,82,-52,-48,-61,-90,95,-49,32,18,40,96,65,44,81,24,38,-4,-14,-97,-20,63,86,-98,-19,-8]
A.sort()
n = len(A) // 2
print(sum(j for j in A[:n]))

-919


**Problema 13.** Se tienen $n$ bloques de longitudes enteras positivas $a_1, a_2, \ldots, a_n$. Se quiere acomodarlos en forma de pirámide, de modo que se obtenga la mayor altura de pirámide posible. ¿Cuál es la altura que se obtiene? Una pirámide es acomodar los bloques en niveles de modo que para cada nivel se tengan más bloques que el nivel de arriba y una suma total mayor que el nivel de arriba. Resolver en general y luego ejecutarlo para: 

```
355,572,788,82,433,225,183,499,374,734,157,121,53,486,954,792,97,684,933,51,868,485,907,292,349,804,83,214,704,200,236,404,276,778,958,105,500,334,448,836,890,537,167,31,398,313,175,930,342
```

Una exploración exhaustiva permutando los bloques acomodándolos seguro que resuelve el problema pero el tiempo de ejecución será muy elevado. Para resolverlo es mejor hacer lo siguiente:

- Ordenamos los bloques de menor a mayor
- Ponemos el menor hasta arriba, los siguientes dos en el nivel inferior, los siguientes tres en el nivel inferior y así sucesivamente
- La altura máxima que podamos completar totalmente es la que se obtiene con este procedimiento

In [4]:
A = [355,572,788,82,433,225,183,499,374,734,157,121,53,486,954,792,97,684,933,51,868,485,907,292,349,804,83,214,704,200,236,404,276,778,958,105,500,334,448,836,890,537,167,31,398,313,175,930,342]
A.sort()

j = 1
suma = 0
piramide = []
suma = 1
while suma < len(A):
    fila = A[suma-j:suma]
    piramide.append(fila)
    j += 1
    suma += j

for f in piramide:
    print(f, sum(f))
        

[31] 31
[51, 53] 104
[82, 83, 97] 262
[105, 121, 157, 167] 550
[175, 183, 200, 214, 225] 997
[236, 276, 292, 313, 334, 342] 1793
[349, 355, 374, 398, 404, 433, 448] 2761
[485, 486, 499, 500, 537, 572, 684, 704] 4467
[734, 778, 788, 792, 804, 836, 868, 890, 907] 7397


**Problema 14.** Se tiene la siguiente pirámide de números. Se comienza desde arriba y en cada paso se puede ir al número inferior, o bien al número inferior a la derecha, hasta que se llega a la fila de hasta abajo. 

## Divide y Conquista

Es partir un problema en problemas más pequeños, que no necesariamente sean instancias del mismo problema. Algunas formas en las que sucede esto es: 

- Partiendo correctamente el problema en funciones auxiliares que sean conceptualmente más fáciles de crear y fáciles de utilizar
- Resolviendo un problema para una familia de instancias y luego usar estas instancias para resolver las demás
- Resolver el problema por casos
- Resolver el problema de manera recursiva.


## Algoritmos recursivos

Es una situación especial de la heurística de divide y conquista, en la cual ponemos a la instancia grande de un problema en términos de instnaicas pequeñas del mismo problema. 

Muchos lenguajes de programación permiten hacer cosas del siguiente estilo:
```
def resolver(instancia grande):
    decir qué hacer si estamos en el caso base
    hacer cosas para obtener instancias pequeñas
    resolver(instancia pequeña)
    hacer cosas
    regresar resultado
```

Aquí la función ```resolver``` se cita a sí misma, de manera muy parecida a como las sucesiones recursiva, como la de Fibonacci, están en términos de ellas mismas. Para que el algoritmo esté bien definido, la función debe decir que hacer con uno (o más) casos base. Veamos un par de ejempolos muy sencillos en Python: 

- Uno en donde definimos la función de "multiplicación por $3$" en términos de la función suma para los números naturales. 
- Otro en donde definimos la función factorial

In [8]:
def multi_3(n):
    if n == 0:
        return 0
    else:
        anterior = multi_3(n-1)
        return 3 + anterior
print(multi_3(10))

def factorial(n):
    if n == 1:
        return 1
    else:
        return n*factorial(n-1)

print(factorial(3))

30
6


**Problema 15.** Escribe un algoritmo recursivo que cree una lista de todas las permutaciones de $n$ elementos. 

Hagamos una lista de todas las permutaciones para pocos elementos:

$n=1$ -> $1$

$n=2$ -> $12,21$

$n=3$ -> $123,132,213,231,312,321$

Esbozo:
- Cuando tenemos un elemento, regresamos el elemento. 
- Cuando tenemos $n$ elementos, primero hacemos las permutaciones de $n-1$ elementos y luego en cada una de ellas ponemos el $n-ésimo$ elemento en cada posición.

In [21]:
def permutaciones(lista):
    if len(lista) == 1:
        return ([lista])
    else:
        anteriores = permutaciones(lista[:-1])
        ultimo = lista[-1]
        nuevas = []
        for anterior in anteriores:
            for j in range(len(anterior) + 1):
                nueva = anterior[:j] + [ultimo] + anterior[j:]
                nuevas.append(nueva)
        return nuevas

perms = permutaciones(['manzana', 'naranja', 'pera', 'uva'])
print(len(perms))
for p in perms:
    print(p)

24
['uva', 'pera', 'naranja', 'manzana']
['pera', 'uva', 'naranja', 'manzana']
['pera', 'naranja', 'uva', 'manzana']
['pera', 'naranja', 'manzana', 'uva']
['uva', 'naranja', 'pera', 'manzana']
['naranja', 'uva', 'pera', 'manzana']
['naranja', 'pera', 'uva', 'manzana']
['naranja', 'pera', 'manzana', 'uva']
['uva', 'naranja', 'manzana', 'pera']
['naranja', 'uva', 'manzana', 'pera']
['naranja', 'manzana', 'uva', 'pera']
['naranja', 'manzana', 'pera', 'uva']
['uva', 'pera', 'manzana', 'naranja']
['pera', 'uva', 'manzana', 'naranja']
['pera', 'manzana', 'uva', 'naranja']
['pera', 'manzana', 'naranja', 'uva']
['uva', 'manzana', 'pera', 'naranja']
['manzana', 'uva', 'pera', 'naranja']
['manzana', 'pera', 'uva', 'naranja']
['manzana', 'pera', 'naranja', 'uva']
['uva', 'manzana', 'naranja', 'pera']
['manzana', 'uva', 'naranja', 'pera']
['manzana', 'naranja', 'uva', 'pera']
['manzana', 'naranja', 'pera', 'uva']


**Problema 16.** Escribe un algoritmo recursivo que decida si una palabra es palíndroma o no, es decir, que diga si una palabra se lee igual al derecho que al revés. 

Para resolverlo en términos recursivos debemos pensar en los casos base que tenemos. Pensaremos en que el caso base consiste en las palabras de o bien $1$ caracter o bien ningún caracter. En ambos casos, estas palabras son palíndromas. 

Si tenemos una palabra más larga, es decir, con tres o más caracteres, una cosa que podemos hacer es comparar el primero con el último y ver si son iguales o no. Si no son iguales, entonces la palabra ya no fue palíndroma. Si sí son iguales, entonces el problema se reduce a que la palabra quitando el primero y el último sea palíndroma. 

In [26]:
def palindroma(palabra):
    if len(palabra) <= 1:
        return True
    if palabra[0] != palabra[-1]:
        return False
    else:
        return palindroma(palabra[1:-1])

print(palindroma('reconocer'))
print(palindroma('alegría'))
print(palindroma('a'))
print(palindroma(''))
print(palindroma('tassat'))
print(palindroma('tatsat'))

True
False
True
True
True
False


## Teorema maestro

## Backtrack y búsquedas combinatorias

Lo que hace una búsqueda combinatoria es recorrer todo el espacio de estados de un problema algorítmico de una forma ordenada. Una forma muy común de hacer esto es mediante un procedimienot recursivo llamado **backtrack**, que progresivamente va construyendo un vector $A=(v_1, v_2, \ldots, v_k)$ mediante la adición o eliminación en la parte final de elementos de una lista de $C$ candidatos.

Cada que el algoritmo llega a un vector que podría ser una solución, lo procesa. 

Para ir recorriendo los vectores, se hace lo siguiente: 
```
backtrack(a, k, datos):
    si a es vacío
        proceso_inicial(a, k, datos)
    si a es válido
        procesar(a, k, datos)
    C = generar_candidatos(a,k,datos)
    para j en C:
        agregar j al final de a
        backtrack(a,k,datos)
        eliminar j del final de a
    proceso_final(a,k,datos)
```

Como hay diferentes valores de $C$ en distintos valores de la recursión, estos no interfieren entre sí. 

### Generar permutacoines con backtrack

In [28]:
def perms(n, a=[], candidates = [], sols = 0):
    if len(a) == 0:
        candidates = list(range(n))
    if len(a) == n:
        sols += 1
        print(a)
        return sols 
    for j in candidates:
        a.append(j)
        new_candidates = candidates.copy()
        new_candidates.remove(j)
        sols = perms(n, a, new_candidates, sols)
        a.remove(j)
    return sols

sols = perms(4)
print(sols)

[0, 1, 2, 3]
[0, 1, 3, 2]
[0, 2, 1, 3]
[0, 2, 3, 1]
[0, 3, 1, 2]
[0, 3, 2, 1]
[1, 0, 2, 3]
[1, 0, 3, 2]
[1, 2, 0, 3]
[1, 2, 3, 0]
[1, 3, 0, 2]
[1, 3, 2, 0]
[2, 0, 1, 3]
[2, 0, 3, 1]
[2, 1, 0, 3]
[2, 1, 3, 0]
[2, 3, 0, 1]
[2, 3, 1, 0]
[3, 0, 1, 2]
[3, 0, 2, 1]
[3, 1, 0, 2]
[3, 1, 2, 0]
[3, 2, 0, 1]
[3, 2, 1, 0]
24


Reto. Utilizar esta idea de backtrack para resolver el problema del reloj que tenemos pendiente. 

In [3]:
def reloj(n=12, a=[], candidates=[], sols=0):
    if len(a) == 0:
        candidates = [1]
    if len(a) == 1:
        candidates = list(range(2,n+1))
    if len(a) == 2:
        print(a[-1])
    if len(a) >= 3:
        if a[-3] + a[-2] + a[-1] == 13:
            return sols 
        if len(a) == n:
            if a[-2] + a[-1] + a[0] == 13 or a[-1] + a[0] + a[1] == 13:
                return sols 
            else:
                return sols + 1
    for j in candidates:
        a.append(j)
        new_candidates = candidates.copy()
        new_candidates.remove(j)
        sols=reloj(n,a,new_candidates,sols)
        a.remove(j)
    return sols

if __name__ == '__main__':
    print(reloj())

2
3
4
5
6
7
8
9
10
11
12
24424416


In [4]:
print(reloj(10))

2
3
4
5
6
7
8
9
10
165520


In [7]:
def reloj_buenos(n=6, a=[], candidates=[], sols=0, buenos=list([])):
    if len(a) == 0:
        candidates = [1]
    if len(a) == 1:
        candidates = list(range(2,n+1))
    if len(a) == 2:
        print(a[-1])
    if len(a) >= 3:
        if a[-3] + a[-2] + a[-1] == 13:
            return sols, buenos
        if len(a) == n:
            if a[-2] + a[-1] + a[0] == 13 or a[-1] + a[0] + a[1] == 13:
                return sols, buenos
            else:
                buenos.append(a.copy())
                return (sols + 1,buenos)
    for j in candidates:
        a.append(j)
        new_candidates = candidates.copy()
        new_candidates.remove(j)
        sols, buenos = reloj_buenos(n,a,new_candidates,sols,buenos)
        a.remove(j)
    return (sols,buenos)

if __name__ == '__main__':
    print(reloj_buenos())

2
3
4
5
6
(56, [[1, 2, 3, 4, 5, 6], [1, 2, 3, 5, 4, 6], [1, 2, 3, 5, 6, 4], [1, 2, 3, 6, 5, 4], [1, 2, 4, 3, 5, 6], [1, 2, 4, 5, 3, 6], [1, 2, 4, 5, 6, 3], [1, 2, 4, 6, 5, 3], [1, 2, 6, 3, 5, 4], [1, 2, 6, 4, 5, 3], [1, 3, 2, 4, 5, 6], [1, 3, 2, 4, 6, 5], [1, 3, 2, 5, 4, 6], [1, 3, 2, 6, 4, 5], [1, 3, 5, 2, 4, 6], [1, 3, 5, 4, 2, 6], [1, 3, 5, 4, 6, 2], [1, 3, 5, 6, 4, 2], [1, 3, 6, 2, 4, 5], [1, 3, 6, 5, 4, 2], [1, 4, 2, 3, 5, 6], [1, 4, 2, 3, 6, 5], [1, 4, 2, 5, 3, 6], [1, 4, 2, 6, 3, 5], [1, 4, 5, 2, 3, 6], [1, 4, 5, 3, 2, 6], [1, 4, 5, 3, 6, 2], [1, 4, 5, 6, 3, 2], [1, 4, 6, 2, 3, 5], [1, 4, 6, 5, 3, 2], [1, 5, 3, 2, 4, 6], [1, 5, 3, 2, 6, 4], [1, 5, 3, 4, 2, 6], [1, 5, 3, 6, 2, 4], [1, 5, 4, 2, 3, 6], [1, 5, 4, 2, 6, 3], [1, 5, 4, 3, 2, 6], [1, 5, 4, 6, 2, 3], [1, 5, 6, 3, 2, 4], [1, 5, 6, 4, 2, 3], [1, 6, 2, 3, 4, 5], [1, 6, 2, 3, 5, 4], [1, 6, 2, 4, 3, 5], [1, 6, 2, 4, 5, 3], [1, 6, 3, 2, 4, 5], [1, 6, 3, 2, 5, 4], [1, 6, 3, 5, 2, 4], [1, 6, 3, 5, 4, 2], [1, 6, 4, 2, 3, 5], [1, 

## Caminos en gráficas

In [1]:
import networkx as nx
G = nx.petersen_graph()